lib/tutf8e/src/tutf8e.c

```diff
@@ -3,121 +3,200 @@
+    break;
+
+  case 'i':
+    if(!strcmp(string_flag,"ignore")) {
+      return TUTF8E_FLAG_INV_IGNORE;
+    }
+    break;
+
+  case 'k':
+    if(!strcmp(string_flag,"keep")) {
+      return TUTF8E_FLAG_INV_KEEP;
+    }
+    break;
+  
+  case 'r':
+    if(!strcmp(string_flag,"replacement")) {
+      return TUTF8E_FLAG_INV_REPLACEMENT;
+    }
+    break;
+  
+  case 'q':
+    if(!strcmp(string_flag,"question")) {
+      return TUTF8E_FLAG_INV_QUESTION;
+    }
+    break;
+  default:
+    break;
+  }
+  return (uint32_t)(-1);
+}
+
 /* Determine the input length and UTF8 encoded length of NUL-terminated input string */
-/* return TUTF8E_INVALID if input character is not convertable                       */
-/* return TUTF8E_OK for success                                                      */
+/* return TUTF8E_INVALID if input character is not convertable  TUTF8E_FLAG_INV_FAIL */
+/* return TUTF8E_SAME    if string doesnt' change  (but copied) */
+/* return TUTF8E_OK      if string changes   */
 
-int tutf8e_string_length(const uint16_t *table, const char *input, size_t *ilen, size_t *olen)
+
+int tutf8e_string_length(const TUTF8encoder encoder, const char *input, size_t *ilen, size_t *olen, uint32_t flags)
 {
+  const uint16_t *table = (uint16_t *) encoder;
   const unsigned char *i;
-  for (i = (const unsigned char *) input; *i; ++i, (*ilen)++) {
+  int changed = 0;
+  int len = 0;
+  for (i = (const unsigned char *) input; *i; ++i) {
     const uint16_t c = table[*i];
+
     if (c<0x80) {
-      *olen += 1;
+      len++;
+      if(c != *i) changed++;
       continue;
     }
+    changed++;
     if (c<0x800) {
-      *olen += 2;
+      len += 2;
       continue;
     }
     if (c<0xffff) {
-      *olen += 3;
+      len += 3;
       continue;
     }
-    return TUTF8E_INVALID;
+    switch(flags & TUTF8E_FLAG_INV_MASK) {
+    case TUTF8E_FLAG_INV_KEEP        : len += *i <= 0x80 ? 1 : 2; break;
+    case TUTF8E_FLAG_INV_FAIL        : return  TUTF8E_INVALID;
+    case TUTF8E_FLAG_INV_IGNORE      : break;
+    case TUTF8E_FLAG_INV_REPLACEMENT : len += 3; break;
+    case TUTF8E_FLAG_INV_QUESTION    : len++; break;
+    case TUTF8E_FLAG_INV_COPY        : len++; break;
+    }
   }
-  return TUTF8E_OK;
+  *ilen = (char*)i- (char *)input;
+  *olen = len;
+  return changed ? TUTF8E_OK : TUTF8E_SAME;
 }
 
+
 /* UTF8 encode the given input string and table                */
 /* olen input is output buffer size, output is encoded length  */
-/* return TUTF8E_TOOLONG if output buffer insuficient          */
+
 /* return TUTF8E_INVALID if input character is not convertable */
 /* return TUTF8E_OK for success                                */
+/* return TUTF8E_SAME : ...                                    */
 
-int tutf8e_string_encode(const uint16_t *table, const char *i, char *o, size_t *olen)
+int tutf8e_string_encode(const TUTF8encoder encoder, const char *input, char *o, uint32_t flags)
 {
   int ret;
-  size_t ilen = 0;
+  size_t ilen = strlen(input);
```

Comment:
Now we're doing _three_ passes of the input, rather than two!  I hope the input is not too long...

Reply:
hum.. that was left from some changes... to be revmoed.